docs2/site/docs/migrations/migration8.md

```diff
@@ -246,6 +246,48 @@ At this time GraphQL.NET does not directly support the `MaxLength` and similar a
 implement your own attributes as shown above, or call the `Validate` method to set a validation
 function.
 
+### 6. `@pattern` custom directive added for validating input values against a regular expression pattern
+
+This directive allows for specifying a regular expression pattern to validate the input value.
+It can also be used as sample code for designing new custom directives, and is now the preferred
+design over the older `InputFieldsAndArgumentsOfCorrectLength` validation rule.
```

Comment:
Why is this preferred over validation rules?

Reply:
Found the answer here
https://github.com/graphql-dotnet/graphql-dotnet/pull/3873/files#diff-72602e5d1f15e15b9dcb3dc7964b37f622e56e31de537a3494941121d2e4abadR257

Reply:
Well, without #3873 , it's really difficult to read argument values.  All you have access to is the raw AST, before it has hit any of the configured scalars or input objects and been coerced to .NET types.  The existing sample is provided in `InputFieldsAndArgumentsOfCorrectLength`.

But let's say we implement #3873 which was my original concept for v8.  Then at least we can write validation rules much easier that are capable of inspecting coerced argument values for any field.  This in fact doesn't fully meet my original goal of universally validating any input value, but let's say we fix that (probably by extending `IVariableVisitor` to work for both literals and variables post-coercion).  Ok so now we can perform the same task in a validation rule with much less code.

So given all those assumptions, now let's compare to how a schema visitor would function.  The complexity of implementation is probably roughly the same now -- if you take a look at `PatternMatchingVisitor` it's pretty simple.  And I'm guessing that after changing `IVariableVisitor` it could get down to about that simple.

But what happens if we implement as a validation rule?  Every execution has to scan the entire incoming request's AST to match on each field argument and see if there are any directives applied.  Pulling the directives list is a dictionary lookup (we should think about eliminating that for v8), and then scanning the applied directives dictionary (if it exists) for an applied directive is another dictionary lookup.  If it finds a match, another dictionary lookup for the value of the argument(s).  When most input fields or field arguments have no attached rules, it seems like a lot of wasted CPU time.  In this regex sample, it would incur regenerating the regex (or pulling from the static dictionary), another costly operation.  Versus with a schema visitor, the exact code necessary to run the validation is built right into the coercion code (into `Validator`) along with the precompiled dependencies - in this case the compiled regex function.

So it becomes obvious that the schema visitor approach is much more optimized because there is no wasted CPU time on portions of the AST that have no validation rules, and any validation rule for any specific field can be 'precompiled' directly into the validation function with no dictionary lookups during runtime.

Reply:
Got it. The code snippet is very useful for understanding.